[Protocol Change] Add the Redirection spec change proposal (#3705)
delta-io/delta: protocol_rfcs/redirection.md

> # Redirection

Reviewer: Please update the table of proposed RFCs - https://github.com/delta-io/delta/tree/master/protocol_rfcs#table-of-rfcs

in fact, i realized that the instructions in that protocol_rfcs/README does not say to update the table. Can you please add that.

Author: Done

**Associated Github issue for discussions: https://github.com/delta-io/delta/issues/3702**
<!-- Remove this: Replace XXXX with the actual github issue number -->
<!--  Give a general description / context of the protocol change, and remove this comment. -->
This RFC proposes a new kind of features called "redirection" for Delta tables. The feature enables seamless migration of existing Delta tables to different storage locations while minimizing disruption to user workflows. This enhancement aims to reduce downtime and simplify the process of relocating Delta tables.
## Terminology:
* Modern Delta clients: These are Delta clients that are compatible with the latest Delta protocol and support the redirection feature.
* Legacy Delta clients: These refer to Delta clients that do not support the redirection feature, as they are not up-to-date with the latest Delta protocol.

> * Redirect Source Table: The source Delta table that is redirected to a new storage location.

Author: Done

* Redirect Destination Table: The Delta table that is created as a redirected table of an existing source table. Any access to the redirect source table would be dispatched to this redirect destination table.
## Overview
The proposal includes two distinct table features:

> * **RedirectReaderWriter**: This feature enables both read and write operations to be redirected from the source to the destination for Modern Delta clients. It blocks all read and write operations from Legacy Delta clients.

Reviewer: can we we remove this term "modern" from everywhere? really ambiguous term... what modern compared to what?

Author: Keep the modern for now as it is described in the terminology section above.

@tdas if there is an alternative name for Modern Delta Client, I would replace it.

### Feature Definition
The Redirection feature provides a redirect type for the Modern Delta clients to parse all other attributes and utilizes the **state** to handle incoming queries.

* **Type**: This is the identifier of the redirection. The Modern Delta clients utilize this value to determine how to parse the redirect specification.

* **State**: This value indicates the status of the redirect table. It has three possible values: **ENABLE-REDIRECT-IN-PROGRESS**, **READY** and **DROP-REDIRECT-IN-PROGRESS**.
	* **ENABLE-REDIRECT-IN-PROGRESS**: This state indicates that the redirect process is still going on. The modern delta client can still read the table on the redirect source table but can't write data to the table. This state also accepts the transaction that updates the value of "delta.redirectReaderWriter" and "delta.redirectWriterOnly".
	* **READY**: This state indicates that the redirect process is completed and the modern delta client can read and write the table. Both table-based and path-based read and write statements are redirected to the destination table.

> 	* **DROP-REDIRECT-IN-PROGRESS**: The table redirection is under withdrawal and the redirection property is going to be removed from the delta table. In this state, the modern delta client stops redirecting new queries to redirect destination tables, and only accepts read-only queries to access the redirect source table. The on-going redirected write or metadata transactions, which are visiting redirect destinations, can not commit. This state accepts the transaction that updates the value of "delta.redirectReaderWriter" and "delta.redirectWriterOnly".

Author: Done

> 	* Concurrent write transactions are aborted using the existing conflict checker mechanism. Therefore, state transitions should commit to both the redirect source and destination tables. This ensures that the Delta table's conflict checker can automatically manage concurrent write transactions.

Reviewer 2: > state transitions should commit to both the redirect source and destination tables. 

Does this mean we are adding the table features to both? Is this documented somewhere else?

Author: >Does this mean we are adding the table features to both?
Ans: Yes

Author: > Is this documented somewhere else?
Ans: Done

> 	* Concurrent write transactions are aborted using the existing conflict checker mechanism. Therefore, state transitions should commit to both the redirect source and destination tables. This ensures that the Delta table's conflict checker can automatically manage concurrent write transactions.

Author: Done

	* Unity Catalog Table:
		* Table: It is a JSON to get the table information of the redirected table. This attribute's value type depends on the Type above. It can be UUID, integer or <database>.<schema>.<name> triple.
```
	delta.redirectReaderWriter: "{
		"Type": "Unity Catalog",
		"State": "READY",
		"Spec": "{
			"Table": "{
				"MetastoreId": "c535492e-375f-4b34-887c-f309ed83bf8d",
				"TableId": "fba84a4d-8789-44db-ae04-5203fbed599b"
			}",
		}",
	}"
```
	* Object Store Table Without Catalog:
		* Location: The URI of the object store path that contains the delta log files of the redirect destination table.
```
	delta.redirectReaderWriter: "{
		"Type": "Object Store",
		"State": "READY",
		"Spec": "{
			"Location": "s3://........./"
		}"
	}"
```

* **NoRedirectRules**: This attribute contains a list of rules for allowing transactions on the redirect source table. It allows maintenance workloads to run on the redirect source tables, for instance, the REFRESH UNIFORM and VACUUM command on the redirect source table. Each rule includes two attributes:

> 	* **AppName**: This is the name of the applications that are allowed to execute commands on the redirect source table. When the sessions with these AppName try to access the redirected table, the modern delta client disables table redirection and their queries would be run on the redirect source table.

Reviewer 2: ```suggestion
	* **AppName**: This is the name of the applications that are allowed to execute commands on the redirect source table. When a session with this AppName tries to access the redirected table, the modern delta client disables table redirection and their queries would be run on the redirect source table.
```

Author: Done
